omp/matrix/coo_kernels.cpp

```diff
@@ -170,33 +255,109 @@ void advanced_spmv2(std::shared_ptr<const OmpExecutor> exec,
-                const auto row = first;
-                const auto col = coo_col[nz];
+            std::array<ValueType, num_rhs> partial_sum;
+            if (first != sentinel_row) {
+                // handle row overlap with previous thread: partial sums
+                partial_sum.fill(zero<ValueType>());
+                for (; nz < end && coo_row[nz] == first; nz++) {
+                    const auto row = first;
+                    const auto col = coo_col[nz];
+#pragma unroll
+                    for (size_type rhs = 0; rhs < num_rhs; rhs++) {
+                        partial_sum[rhs] +=
+                            scale * coo_val[nz] * b->at(col, rhs);
+                    }
+                }
+                // handle row overlap with previous thread: add to memory
+#pragma unroll
                 for (size_type rhs = 0; rhs < num_rhs; rhs++) {
-                    atomic_add(c->at(row, rhs),
-                               scale * coo_val[nz] * b->at(col, rhs));
+                    atomic_add(c->at(first, rhs), partial_sum[rhs]);
                 }
             }
+            // handle non-overlapping rows
             for (; nz < end && coo_row[nz] != last; nz++) {
                 const auto row = coo_row[nz];
                 const auto col = coo_col[nz];
+#pragma unroll
                 for (size_type rhs = 0; rhs < num_rhs; rhs++) {
                     c->at(row, rhs) += scale * coo_val[nz] * b->at(col, rhs);
                 }
             }
-            for (; nz < end; nz++) {
-                const auto row = last;
-                const auto col = coo_col[nz];
+            if (last != sentinel_row) {
+                // handle row overlap with following thread: partial sums
+                partial_sum.fill(zero<ValueType>());
+                for (; nz < end; nz++) {
+                    const auto row = last;
+                    const auto col = coo_col[nz];
+#pragma unroll
+                    for (size_type rhs = 0; rhs < num_rhs; rhs++) {
+                        partial_sum[rhs] +=
+                            scale * coo_val[nz] * b->at(col, rhs);
+                    }
+                }
+                // handle row overlap with following thread: add to memory
+#pragma unroll
                 for (size_type rhs = 0; rhs < num_rhs; rhs++) {
-                    atomic_add(c->at(row, rhs),
-                               scale * coo_val[nz] * b->at(col, rhs));
+                    const auto row = last;
+                    atomic_add(c->at(last, rhs), partial_sum[rhs]);
                 }
             }
         }
     }
 }
 
+
+template <typename ValueType, typename IndexType>
+void generic_spmv2(std::shared_ptr<const OmpExecutor> exec,
+                   const matrix::Coo<ValueType, IndexType>* a,
+                   const matrix::Dense<ValueType>* b,
+                   matrix::Dense<ValueType>* c, ValueType scale)
+{
+    const auto num_rhs = b->get_size()[1];
+    if (num_rhs <= 0) {
```

Comment:
How do you motivate this choice of dispatches? Did you test, if `spmv2_small_rhs` is slower than `spmv2_blocked` for more than 4 rhs? I'm not against this choice, just curious.